odoo/addons/base/models/ir_ui_menu.py

```diff
@@ -74,62 +74,69 @@ def _check_parent_id(self):
-        existing_actions = {
-            action
-            for model_name, action_ids in actions_by_model.items()
-            for action in self.env[model_name].browse(action_ids).exists()
-        }
-        action_menus = menus.filtered(lambda m: m.action and m.action in existing_actions)
-        folder_menus = menus - action_menus
-        visible = self.browse()
+                action_ids_by_model[action._name].append(action.id)
 
-        # process action menus, check whether their action is allowed
-        access = self.env['ir.model.access']
         MODEL_BY_TYPE = {
             'ir.actions.act_window': 'res_model',
             'ir.actions.report': 'model',
             'ir.actions.server': 'model_name',
         }
+        def exists_actions(model_name, action_ids):
+            """ Return existing actions and fetch model name field if exists"""
+            if model_name not in MODEL_BY_TYPE:
+                return self.env[model_name].browse(action_ids).exists()
+            records = self.env[model_name].sudo().with_context(active_test=False).search_fetch(
+                [('id', 'in', action_ids)], [MODEL_BY_TYPE[model_name]], order='id',
+            )
+            if model_name == 'ir.actions.server':
+                # Because it is computed, `search_fetch` doesn't fill the cache for it
+                records.mapped('model_name')
+            return records
 
-        # performance trick: determine the ids to prefetch by type
-        prefetch_ids = defaultdict(list)
-        for action in action_menus.mapped('action'):
-            prefetch_ids[action._name].append(action.id)
+        existing_actions = {
+            action
+            for model_name, action_ids in action_ids_by_model.items()
+            for action in exists_actions(model_name, action_ids)
+        }
+        action_menus = menus.filtered(lambda m: m.action in existing_actions)
+        folder_menu_ids = set((menus - action_menus)._ids)
 
+        # process action menus, check whether their action is allowed
+        access = self.env['ir.model.access']
+        visible_ids = set()
         for menu in action_menus:
             action = menu.action
-            action = action.with_prefetch(prefetch_ids[action._name])
+            # action[MODEL_BY_TYPE[action._name]] has been fetched in `exists_actions`
             model_name = action._name in MODEL_BY_TYPE and action[MODEL_BY_TYPE[action._name]]
             if not model_name or access.check(model_name, 'read', False):
                 # make menu visible, and its folder ancestors, too
-                visible += menu
+                visible_ids.add(menu.id)
                 menu = menu.parent_id
-                while menu and menu in folder_menus and menu not in visible:
-                    visible += menu
+                while menu.id not in visible_ids and menu.id in folder_menu_ids:
+                    visible_ids.add(menu.id)
                     menu = menu.parent_id
```

Comment:
I'm not sure why you filter on `folder_menu_ids`, I think it is easier to understand filtering on all menus.
```python
# make menu visible, and its folder ancestors, too
while menu in menus:  # TODO just make a set of ids
    visible_ids.add(menu.id)
    menu = menu.parent_id
```

Reply:
Yep, you are right. I updated the code and the banchmark (1ms faster :1st_place_medal: ).